apps/dcos_net/src/dcos_net_mesos_listener.erl

```diff
@@ -986,14 +1027,108 @@ stream(Data, #state{pid=Pid, size=Size, buf=Buf}=State) ->
+    try
+        prometheus_counter:inc(mesos_listener, messages_total, [], 1)
+    catch error:_Error ->
+        ok
+    end,
+    try jiffy:decode(Head, [return_maps]) of Obj ->
+        {next, Obj, State0}
+    catch error:Error ->
+        {error, Error}
+    end.
+
+%%%===================================================================
+%%% Metrics functions
+%%%===================================================================
+
+-spec(init_metrics() -> ok).
+init_metrics() ->
+    init_metrics_pubsub(),
+    init_metrics_mesos_state(),
+    init_metrics_received().
+
+init_metrics_pubsub() ->
+    prometheus_summary:new([
+        {registry, mesos_listener},
+        {name, pubsub_duration_seconds},
+        {labels, []},
+        {help, "Time spent processing pubsub"}]).
+
+init_metrics_received() ->
+    prometheus_counter:new([
+        {registry, mesos_listener},
+        {name, bytes_total},
+        {labels, []},
+        {help, "Total number of bytes received"}]),
+    prometheus_counter:new([
+        {registry, mesos_listener},
+        {name, failures_total},
+        {labels, []},
+        {help, "Total number of failures"}]),
+    prometheus_boolean:new([
+        {registry, mesos_listener},
+        {name, listening_leader_boolean},
+        {labels, []},
+        {help, "Boolean indicating that listening is from a mesos leader. " ++
+            " 0 indicates standby master, 1 indicates leader."}]),
+    prometheus_counter:new([
+        {registry, mesos_listener},
+        {name, messages_total},
+        {labels, []},
+        {help, "Total number of messages"}]).
+
+init_metrics_mesos_state() ->
+    prometheus_gauge:new([
+        {registry, mesos_listener},
+        {name, agents_total},
+        {labels, []},
+        {help, "Total number of Agents"}]),
+    prometheus_gauge:new([
+        {registry, mesos_listener},
+        {name, frameworks_total},
+        {labels, []},
+        {help, "Total number of Frameworks"}]),
+    prometheus_gauge:new([
+        {registry, mesos_listener},
+        {name, tasks_total},
+        {labels, []},
+        {help, "Total number of Tasks"}]),
+    prometheus_gauge:new([
+        {registry, mesos_listener},
+        {name, waiting_tasks_total},
+        {labels, []},
+        {help, "Total number of tasks with no agent/framework information"}]).
+
+-spec(handle_metrics(state()) -> state()).
+handle_metrics(#state{agents=A, frameworks=F,
+        tasks=T, waiting_tasks=WT}=State) ->
+    try
+        prometheus_gauge:set(
+            mesos_listener, tasks_total, [],
+            maps:size(T)),
+        prometheus_gauge:set(
+            mesos_listener, waiting_tasks_total, [],
+            maps:size(WT)),
+        prometheus_gauge:set(
+            mesos_listener, frameworks_total, [],
+            maps:size(F)),
+        prometheus_gauge:set(
+            mesos_listener, agents_total, [],
+            maps:size(A))
+    catch error:_Error ->
+        ok
+    end,
```

Comment:
You actually don't need `try catch end` wrappers in this module for metrics. There is only one test that uses this module directly and it tests only this module. So it makes sense to fix the eunit test instead of adding wrappers here and, possibly, making the code less robust. It's a rare case when you have to use `try catch end` wrappers for metrics, e.g. a lot of tests that are calling the code, but not checking the results (dns zones) or other applications are incrementing counters (dns zones, logger metrics). It just works with a small patch:

```
diff --git a/apps/dcos_net/test/dcos_net_mesos_listener_tests.erl b/apps/dcos_net/test/dcos_net_mesos_listener_tests.erl
index 8ec6dd2..9e1fdb7 100644
--- a/apps/dcos_net/test/dcos_net_mesos_listener_tests.erl
+++ b/apps/dcos_net/test/dcos_net_mesos_listener_tests.erl
@@ -553,6 +553,9 @@ setup(FileName) ->
             Pid ! stream_next
         end),
 
+    ok = application:start(prometheus),
+    dcos_net_mesos_listener:init_metrics(),
+
     {ok, _Pid} = dcos_net_mesos_listener:start_link(),
     stream_wait(),
 
@@ -569,6 +572,8 @@ cleanup(_Tasks) ->
     exit(Pid, kill),
     exit(StreamPid, kill),
 
+    ok = application:stop(prometheus),
+
     meck:unload(httpc).
 
 stream_wait() ->
```